File: 백준/Platinum/1637. 날카로운 눈/날카로운 눈.py
import sys
from collections import Counter
input = lambda: sys.stdin.readline().strip()

dummies = [list(map(int,input().split())) for _ in range(int(input()))]

def search(n):
    result = 0
    for a,b,c in dummies:
        if a>n or a>b:
            continue
        result +=(min(b,n)-a)//c+1
    return result

# print(search(1))
s,e = 1, 2_147_483_647

if search(e)%2==0:
    print('NOTHING')
    sys.exit()
result = e
while s<=e:
    m = (s+e)//2
    v = search(m)
    # print(s,m,e)
    if v%2:
        e = m-1
        result = min(result, m)
    else:
        s = m+1
print(result, search(result)-search(result-1))

File: 백준/Platinum/1637. 날카로운 눈/날카로운 눈.java
import java.io.*;
import java.util.*;

public class Main {
    static List<int[]> dummies = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            dummies.add(new int[]{a, c, b});
        }

        long s = 1, e = 2_147_483_647L;

        if (search(e) % 2 == 0) {
            System.out.println("NOTHING");
            return;
        }

        long result = e;
        while (s <= e) {
            long m = (s + e) / 2;
            long v = search(m);
            
            if (v % 2 == 1) {
                e = m - 1;
                result = Math.min(result, m);
            } else {
                s = m + 1;
            }
        }

        System.out.println(result + " " + (search(result) - search(result - 1)));
    }

    static long search(long n) {
        long count = 0;
        for (int[] dummy : dummies) {
            int a = dummy[0], c = dummy[1], b = dummy[2];
            if (a > n) continue;
            count += (Math.min(c, n) - a) / b + 1;
        }
        return count;
    }
}